cw_nano/lib/nano_wallet_service.dart

```diff
@@ -0,0 +1,160 @@
+  NanoWalletService(this.walletInfoSource);
+
+  final Box<WalletInfo> walletInfoSource;
+
+  static bool walletFilesExist(String path) =>
+      !File(path).existsSync() && !File('$path.keys').existsSync();
+
+  @override
+  WalletType getType() => WalletType.nano;
+
+  @override
+  Future<WalletBase> create(NanoNewWalletCredentials credentials) async {
+    // nano standard:
+    DerivationType derivationType = DerivationType.nano;
+    String seedKey = NanoSeeds.generateSeed();
+    String mnemonic = NanoUtil.seedToMnemonic(seedKey);
+
+    // bip39:
+    // derivationType derivationType = DerivationType.bip39;
+    // String mnemonic = bip39.generateMnemonic();
+
+    credentials.walletInfo!.derivationType = derivationType;
+
+    final wallet = NanoWallet(
+      walletInfo: credentials.walletInfo!,
+      mnemonic: mnemonic,
+      password: credentials.password!,
+    );
+    wallet.init();
+    return wallet;
+  }
+
+  @override
+  Future<void> remove(String wallet) async {
+    final path = await pathForWalletDir(name: wallet, type: getType());
+    final file = Directory(path);
+    final isExist = file.existsSync();
+
+    if (isExist) {
+      await file.delete(recursive: true);
+    }
+
+    final walletInfo = walletInfoSource.values
+        .firstWhere((info) => info.id == WalletBase.idFor(wallet, getType()));
+    await walletInfoSource.delete(walletInfo.key);
+  }
+
+  @override
+  Future<void> rename(String currentName, String password, String newName) async {
+    final currentWalletInfo = walletInfoSource.values
+        .firstWhere((info) => info.id == WalletBase.idFor(currentName, getType()));
+
+    currentWalletInfo.derivationType = DerivationType.nano; // doesn't matter for the rename action
+
+    String randomWords =
+        (List<String>.from(nm.NanoMnemomics.WORDLIST)..shuffle()).take(24).join(' ');
+    final currentWallet =
+        NanoWallet(walletInfo: currentWalletInfo, password: password, mnemonic: randomWords);
+
+    await currentWallet.renameWalletFiles(newName);
+
+    final newWalletInfo = currentWalletInfo;
+    newWalletInfo.id = WalletBase.idFor(newName, getType());
+    newWalletInfo.name = newName;
+
+    await walletInfoSource.put(currentWalletInfo.key, newWalletInfo);
+  }
+
+  @override
+  Future<NanoWallet> restoreFromKeys(NanoRestoreWalletFromKeysCredentials credentials) async {
+    if (credentials.seedKey.contains(' ')) {
+      throw Exception("Invalid key!");
+    } else {
+      if (credentials.seedKey.length != 64 && credentials.seedKey.length != 128) {
+        throw Exception("Invalid key length!");
+      }
+    }
+
+    DerivationType derivationType = credentials.derivationType ?? DerivationType.nano;
+    credentials.walletInfo!.derivationType = derivationType;
+
+    final wallet = await NanoWallet(
+      password: credentials.password!,
+      mnemonic: credentials.seedKey, // we can't derive the mnemonic from the key in all cases
```

Comment:
this case is extremely rare, basically nano-style mnemonics can go from mnemonic to hex and back, but non-nano (bip39) nano mnemonics can only go from mnemonic to hex, but you can't go from hex back to mnemonic

the majority of nano users use nano mnemonics, the only notable exceptions are ledger nano device users who have exported their seed, and trust wallet users
but even in theses cases they would have to manually convert their seed into it's hex form as these applications present them in their mnemonic form

this comment is essentially saying that if a user puts in a bip39 hex seed, we can't show them the mnemonic as going from hex->mnemonic isn't possible (and we should instead show the hex seed)
but again the percent of users this would affect is basically 0

this comment reminds me of a similar edge case I think I forgot to handle though, so I'll add that now